redash/extensions.py

```diff
@@ -1,30 +1,72 @@
+        return False
+
+
+def entry_point_module(entry_point):
+    """Returns the dotted module path for the given entry point"""
+    return entry_point.pattern.match(entry_point.value).group('module')
+
+
+def module_bundle_files(module):
+    if not resource_isdir(module, BUNDLE_DIRECTORY):
+        return
+
+    with path(module, BUNDLE_DIRECTORY) as bundle_dir:
+        # Copy content of extension bundle into extensions directory
+        return list(bundle_dir.rglob("*"))
 
-    for entry_point in iter_entry_points('redash.extensions'):
+
+def init_extensions(app):
+    """Load the Redash extensions for the given Redash Flask app.
+
+    The extension entry point can return any type of value but
+    must take a Flask application object.
+
+    E.g.::
+
+        def extension(app):
+            app.logger.info("Loading the Foobar extenions")
+            Foobar(app)
+
+    """
+    for entry_point in entry_points().get('redash.extensions', []):
         app.logger.info('Loading Redash extension %s.', entry_point.name)
+        module = entry_point_module(entry_point)
+        # First of all, try to get a list of bundle files
+        extensions[entry_point.name]['resource_list'] = module_bundle_files(module)
+
         try:
-            extension = entry_point.load()
-            app.redash_extensions[entry_point.name] = {
-                "entry_function": extension(app),
-                "resources_list": []
-            }
-        except ImportError:
-            app.logger.info('%s does not have a callable and will not be loaded.', entry_point.name)
-            (root_module, _) = os.path.splitext(entry_point.module_name)
-            content_folder_relative = os.path.join(entry_point.name, 'bundle')
-
-            # If it's a frontend extension only, store a list of files in the bundle directory.
-            if resource_isdir(root_module, content_folder_relative):
-                app.redash_extensions[entry_point.name] = {
-                    "entry_function": None,
-                    "resources_list": resource_listdir(root_module, content_folder_relative)
-                }
+            # Then try to load the entry point (import and getattr)
+            obj = entry_point.load()
+        except (ImportError, AttributeError):
+            # or move on
+            app.logger.error('Extension %s could not be found.', entry_point.name)
+            extensions[entry_point.name]['extension'] = None
+            continue
+
+        # Otherwise check if the loaded entry point is a module
+        if isinstance(obj, ModuleType):
+            app.logger.info('Extension %s is a module.', entry_point.name)
+            extensions[entry_point.name]['extension'] = obj
+        # or simply call the loaded entry point instead.
+        else:
+            app.logger.info('Extension %s is a callable.', entry_point.name)
+            extensions[entry_point.name]['extension'] = obj(app)
```

Comment:
Maybe we should always expect a callable? Is there a use for having a module supported?

Reply:
Yeah, technically entry points are allowed to refer to modules, so we'll need to handle it somehow, but we can instead check if the return value is a callable and otherwise throw an exception. 👍

Reply:
Yep, done.